[cutlass backend] Forward fix for less aligned gemm shapes

henrylhtsang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #148521

Differential Revision: [D70600093](https://our.internmc.facebook.com/intern/diff/D70600093/)

1. Check if config name filtering still works.
Tested, it works

2. do we get C++ compile error
Yes, potentially we need to filter them out manually.

Here we get this.
```
static_assert(threads_minor == 0 || (TileSizeK % threads_minor == 0));
```
We need to move some assertions to gemm_template.py

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

henrylhtsang committed:
[cutlass backend] Forward fix for less aligned gemm shapes



Differential Revision: [D70600093](https://our.internmc.facebook.com/intern/diff/D70600093/)

[ghstack-poisoned]

henrylhtsang committed:
Update on "[cutlass backend] Forward fix for less aligned gemm shapes"


Differential Revision: [D70600093](https://our.internmc.facebook.com/intern/diff/D70600093/)

NOTE: Need to check if config name filtering still works.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

ColinPeppler commented on test/inductor/test_cutlass_backend.py:
```diff
@@ -964,6 +965,79 @@ def select_no_algorithm(*args, **kwargs):
+                with self.assertRaises(InductorError):
+                    torch.compile(torch.mm, dynamic=False)(*input)
+
+                self.assertTrue(
+                    sa.called,
+                    f"autotune_select_algorithm was not called  with shape M={M}, N={N}, K={K}",
+                )
+                args, _ = sa.call_args
+                op_name, choices, _, __ = args
+                assert op_name == "mm"
+                cuda_template_count = 0
+                for choice in choices:
+                    if isinstance(choice, CUDATemplateCaller):
+                        choice_info = choice.info_dict()
+                        op_conf_name = choice_info.get("op_conf_name", "")
+                        assert isinstance(op_conf_name, str)
+                        cuda_template_count += 1
+
+                self.assertGreater(
+                    cuda_template_count,
+                    0,
+                    f"No CUDATemplateCaller choices found for matrix multiplication with shape M={M}, N={N}, K={K}",
+                )
```
what are you trying to do here? it seems like you just want to see if there's an available CUTLASS choice?

henrylhtsang replied:
> it seems like you just want to see if there's an available CUTLASS choice for each of the shapes represented by inputs?

Yes

> if so, why not just check that there exists a CUTLASS kernel which could be a simpler check but functionally the same.

what do you mean?

Also I am copying the tests from the allowlist denylist tests

ColinPeppler commented on torch/_inductor/codegen/cuda/gemm_template.py:
```diff
@@ -1212,46 +1214,29 @@ def _set_bias_layout_and_alignment(
-                return False
-        return True
-
-    def _dtype_match(
```
assuming, this isn't used anymore and no plans to re-use it

ColinPeppler commented on torch/_inductor/codegen/cuda/gemm_template.py:
```diff
@@ -852,14 +862,6 @@ def filter_op(
             ):
                 return None
 
-        # Set bias layout and alignment.
-        status = self._set_bias_layout_and_alignment(op)
-        if not status:
-            log.debug(
-                "Skipping due to bias layout and alignment setting failure. op: %s", op
-            )
-            return None
-
```
I see, so in the case where K wasn't divisible by 8 (but divisible by 4). We'd filter out ops due to allowlist or denylist, that would've work in the divisible by 4 case?

henrylhtsang replied:
> I see, so in the case where K wasn't divisible by 8 (but divisible by 4). We'd filter out ops due to allowlist or denylist, that would've work in the divisible by 4 case?

Umm what do you mean?

allowlist and denylist are user configs. Default is None.

I am not filtering the less aligned gemms. It is filtered out via set_alignment, which basically checks if the alignment is small enough. Say K is only divisible by 2 but not 4. Then most ops except the align2 ones will get filtered out there.

What I am doing here is to move the denylist allowlist filtering to the end, since I want to make sure configuration name doesn't change anymore.

ColinPeppler replied:
Gotcha, I was curious about that last part.

henrylhtsang committed:
Update on "[cutlass backend] Forward fix for less aligned gemm shapes"


Differential Revision: [D70600093](https://our.internmc.facebook.com/intern/diff/D70600093/)

NOTE: Need to check if config name filtering still works.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

henrylhtsang commented:
@pytorchbot merge -i

davidberard98 commented:
@pytorchbot revert -c nosignal -m "broke lint: [GH job link](https://github.com/pytorch/pytorch/actions/runs/13690720601/job/38283359447) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/ad49cfc9f0a8a4d8881b3734edd8c33a087c8b97)"

henrylhtsang committed:
Update on "[cutlass backend] Forward fix for less aligned gemm shapes"


Differential Revision: [D70600093](https://our.internmc.facebook.com/intern/diff/D70600093/)

1. Check if config name filtering still works.
Tested, it works

2. do we get C++ compile error
Yes, potentially we need to filter them out manually.

Here we get this.
```
static_assert(threads_minor == 0 || (TileSizeK % threads_minor == 0));
```
We need to move some assertions to gemm_template.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov 

[ghstack-poisoned]

henrylhtsang commented:
@pytorchbot merge -i

henrylhtsang commented:
@pytorchbot revert -c nosignal -m "mistakes when writing the tests"

henrylhtsang committed:
Update on "[cutlass backend] Forward fix for less aligned gemm shapes"


Differential Revision: [D70600093](https://our.internmc.facebook.com/intern/diff/D70600093/)

1. Check if config name filtering still works.
Tested, it works

2. do we get C++ compile error
Yes, potentially we need to filter them out manually.

Here we get this.
```
static_assert(threads_minor == 0 || (TileSizeK % threads_minor == 0));
```
We need to move some assertions to gemm_template.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov 

[ghstack-poisoned]

henrylhtsang commented:
@pytorchbot merge -i
